### PR TITLE
`Assessment`: Display subsequent feedback in the assessment review when the usage limit of assessment instruction is exceed

### LINKED ISSUE #4442: Feedback that is not included in total score leads to false impression as a student
### Describe the bug

Manual assessment allows to use structured grading criteria. When the instruction is  applied to feedback more than its usage limit, Artemis does not include the points from from feedback that exceeded associated instruction's usage limit into total score. However, student can not distinguish which feedback included in the total score. It is confusing from the students' perspective. 

### To Reproduce

1. Create an exercise and add assessment instructions
2. Define usage limit for the instruction
3. Participate this exercise as student
4. Assess this submission using the structured assessment instructions
5. Use the instruction more than its usage limit
6. See that the points not included in the total score
7. After the assessment period, login as student and view the submission
8. Calculate the score according to the feedback received, see that it does not match the total score

### Expected behavior

The student should be able to see which feedback is included in the total score

### Screenshots

usage limit is 1 for the assessment instruction
image

The instruction is used 3 times to give feedback, total score is 2 because Artemis calculated the total score considering the usage limits of the assessment instruction. From the student side, total score should be 6, because feedback does not declare if its included in the total score or not.
image

### What browsers are you seeing the problem on?

Chrome

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by imstreetspirit**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ]* This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [ ]* I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ]* I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [ ]* I implemented the changes with a good performance and prevented too many database calls.
- [ ]* I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [ ]* I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ]* I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ]* I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

Fixes #4442 
Manual assessment allows using of structured grading criteria. When the instruction is applied to feedback more than its usage limit, Artemis does not include the points from feedback that exceeded the associated instruction's usage limit into a total score. However, students can not distinguish which feedback is included in the total score. It is confusing from the students' perspective. To fix this issue, we separated the subsequent feedback on the client-side considering the usage limits of the grading criteria. Now, the students can understand which feedback is not part of the assessment. Since Artemis offers grading criteria for all Programming Exercise(manual assessment), Modeling Exercise, Text Exercise,  File-Upload Exercise, we integrated this improvement for all submission types. 

### Description

- `checkSubsequentFeedbackInAssessment` added in the `feedback.model`, it receives all feedback in the assessment and marks the subsequent feedback considering the usage limit of the associated grading instruction.
- submission review templates changed to display the subsequent feedback differently and added an icon with a tooltip message
- after receiving the feedback for the assessment in the submission review components, `checkSubsequentFeedbackInAssessment` is added to find the subsequent feedback in the assessments.
- When the usage limit for the instructions changed during the assessment period, the grading criteria needs to be updated to calculate the results for the re-evaluation. current logic calculates the results according to the outdated usage limits. It is fixed with this PR. 
- I also modified the tooltip message for the grading scale property of the assessment instruction to avoid misunderstanding as discussed [here](https://github.com/ls1intum/Artemis/issues/4515). Feel free to suggest a better translation in German!

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise, 1 Modeling Exercise, 1 Text Exercise, 1 File-Upload Exercise

This improvement is integrated for the exercises that enable grading criteria in the manual assessments. So, we need to test the functionality for each type of exercise(except quizzes). The testing steps are quite similar for different types of exercises, so I won't separate them.

1. Log in to Artemis as Instructor
2. Navigate to Course Management
3. Create 4 exercises: 1 Programming Exercise, 1 Modeling Exercise, 1 Text Exercise, 1 File-Upload Exercise
4. Define Assessment Instructions for the exercises (for programming you need to allow manual assessment)
4a. For example, add two instructions and set the usage limit 0 for one of them, then set the usage limit 1 for the other one.
5. As a student, submit a solution for all exercises you created at the previous step
6. As an instructor, assess all those submissions using the assessment instruction
6a. use the instruction (its usage limit is 0) twice while assessing the submission, using means that drag and drop the instruction to the feedback area
6b. use the instruction (its usage limit is 1) three times while assessing the submission. In this case, Artemis will not include the credit of the instruction into the total score for the second and third usage. 
6c. add additional feedback using the button at the bottom of the page. drag and drop the instruction (its usage limit is 1) for this additional feedback.
7. Submit/Override the assessment.
8. As a student, review your submissions, look at the feedback on the page confirm that the subsequent feedback is displayed differently, you can understand why that feedback is not part of the total score. Look at the screenshots below.

_Additional Testing Steps_ (you can test the following steps for only one exercise type, e.g text exercise)
1. After completing the steps above, as an Instructor, change the usage limit of the instruction.
2. Save the exercise, then you will see a confirmation dialog, choose the "re-evaluate" option. 
3. see that results are calculated according to the new usage limit properly
4. As a student, review your submission again, the number of subsequent feedback should be updated

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests (Text)
- [x] Test 1
- [x] Test 2
#### Manual Tests (Programming)
- [x] Test 1
- [x] Test 2
#### Manual Tests (Modeling)
- [x] Test 1
- [x] Test 2
#### Manual Tests (File-Upload)
- [x] Test 1
- [x] Test 2

### Test Coverage

text-result.component.ts  |   95.38 
modeling-submission.component.ts  |   85.66 

### Screenshots

Modeling Submission
image

Text Submission
image

File Upload Submission
image

Programming Submission
image

### CODE REVIEW

**Review by Strohgelaender — CHANGES_REQUESTED** [review]
Tested on TS1 with @cremertim @VinnavanSembian @lBruchmann @willergeorg @Bofan-Zhu. For Programming, Modeling, and File Upload Exercises everything works as expected. The overused grading instructions are displayed gray for the student and reevaluating after changing the maximum usage correctly updates the score and the changes are reflected in the students view.

But for text exercises, the reevaluation does not work. The re-evaluate endpoints returns an internal server error due to a null pointer exception: "Cannot invoke "de.tum.in.www1.artemis.domain.Result.getId()" because the return value of "de.tum.in.www1.artemis.domain.Submission.getLatestResult()" is null".

Additionally, it would be nice to directly show the gray background while assessing the submission. When the tutor uses a grading instruction more than the defined limit, every usage still has a green background. I think we could prevent assessment mistakes when the tutor already sees the instruction gray here.

**Review by VinnavanSembian — COMMENTED** [review]
Tested on ts1 with @Strohgelaender @lBruchmann @cremertim @willergeorg @Bofan-Zhu. See @Strohgelaender for remarks!

**Review by willergeorg — COMMENTED** [review]
See @Strohgelaender s remarks

**Review by lBruchmann — COMMENTED** [review]
Tested on ts1. I followed the testing steps. Everything works as described. See @Strohgelaender  for some remarks.

**Review by Bofan-Zhu — COMMENTED** [review]
Tested on ts1 with @Strohgelaender @VinnavanSembian @cremertim @lBruchmann and
@willergeorg. Works fine for every exercise type except text exercises -> see @Strohgelaender's request for changes.

**Review by cremertim — COMMENTED** [review]
see remarks from @Strohgelaender

**Review by Strohgelaender — APPROVED** [review]
Retested on TS1, the re-evaluation is fixed and the subsequent feedback is shown correctly.

**Review by Strohgelaender — CHANGES_REQUESTED** [review]
Tested on TS1 with @lBruchmann @willergeorg @VinnavanSembian. The feedback was shown correctly in the feedback popup, but there are some smaller inconsistencies when viewing the own submission. There is also a huge problem with complaints and the grading instructions:

For programming exercises, the code editor still showed every feedback with a green color, even if its subsequent feedback.

grafik

For modeling exercises, the table shows the feedback in gray, but the corresponding element stil has a green tick. 

grafik
(ignore the "Limit: 2"-text here, we re-evaluated it to 1 afterwards, this behavior is correct)

After accepting a complaint, the connection with the instruction apparently gets removed. There is no more gray feedback visible for the students and the points are counted towards the score, leaving the student with a higher score than expected.

**Review by willergeorg — COMMENTED** [review]
Tested during testing session with a programming exercise and a modeling exercise. We observed some issues. See @Strohgelaender s remarks.

**Review by VinnavanSembian — COMMENTED** [review]
tested during testing session with @Strohgelaender @lBruchmann @willergeorg on ts1. See @Strohgelaender for remarks.

**Review by lBruchmann — APPROVED** [review]
Tested on ts1. Tested together with @Strohgelaender, @VinnavanSembian and @willergeorg. See @Strohgelaender for some remarks .

**Comment by imstreetspirit** [comment]
Thanks for your review again.

> For programming exercises, the code editor still showed every feedback with a green color, even if its subsequent feedback.

I think it can be resolved in the scope of this PR.

> For modeling exercises, the table shows the feedback in gray, but the corresponding element still has a green tick.

the model is shown in the Apollon editor, so if we want to change it requires some changes/additional logic on the Apollon side. I am not sure if we want this logic in the Apollon editor, I can ask. However, I do not think that it affects what this PR wants to achieve. 

> After accepting a complaint, the connection with the instruction apparently gets removed. There is no more gray feedback visible for the students and the points are counted towards the score, leaving the student with a higher score than expected.

well, I am not familiar with the newly changed complaint handling process in Artemis and I did not touch any part related to complain logic, it seems like an issue with result calculation after accepting a complaint. It does not relate to this PR, feel free to open a ticket for this issue.

**Comment by imstreetspirit** [comment]
Now the code editor displays the subsequent feedback for students. The PR is ready to review/test again.

File: src/main/webapp/i18n/de/exercise.json [hunk 1]
```diff
@@ -64,7 +64,7 @@
             "editText": "Als Text bearbeiten",
             "usageCountHint": "Diese Zahl gibt an, wie oft Punkte für diese Anweisung zur Endbewertung hinzugefügt werden.",
             "creditsHint": "Diese Zahl gibt die Punktzahl für diese Bewertungsanweisung an.",
-            "gradingScaleHint": "Die Benotungsskala definiert die untere und obere Grenze für die Bewertung des Leistungsniveaus von Studierenden mit dieser Anweisung.",
+            "gradingScaleHint": "Die Skala beschreibt das Leistungsniveau für den Anweisung. Einige Beispiele sind ausgezeichnet, gut, durchschnittlich, schlecht.",
```
Comment (willergeorg) [inline src/main/webapp/i18n/de/exercise.json:67]:
```suggestion
            "gradingScaleHint": "Die Skala beschreibt das Leistungsniveau für den Anweisung. Einige Beispiele hierfür sind: ausgezeichnet, gut, durchschnittlich, schlecht.",
```

File: src/main/webapp/i18n/en/exercise.json [hunk 2]
```diff
@@ -58,7 +58,7 @@
             "limit": "Limit",
             "usageCountHint": "This number specifies how many times the score of this instruction may be included in the final score.",
             "creditsHint": "This number specifies the score for this instruction.",
-            "gradingScaleHint": "The grading scale defines the lower and upper limit for assessing student performance levels with this grading instruction.",
+            "gradingScaleHint": "The scale describes the performance level for the instruction. Some examples are excellent, good, average, poor.",
```
Comment (willergeorg) [inline src/main/webapp/i18n/en/exercise.json:61]:
```suggestion
            "gradingScaleHint": "The scale describes the performance level for the instruction. Some examples are: excellent, good, average, poor.",
```

File: src/test/javascript/spec/component/modeling-submission/modeling-submission.component.spec.ts [hunk 3]
```diff
@@ -428,4 +428,47 @@ describe('ModelingSubmission Management Component', () => {
...
+            } as Feedback,
+        ];
+
+        comp.assessmentResult.feedbacks = feedbacks;
+
+        const unreferencedFeedback = comp.unreferencedFeedback;
+        const referencedFeedback = comp.referencedFeedback;
+
+        expect(unreferencedFeedback).not.toBe(undefined);
+        expect(unreferencedFeedback!.length).toBe(1);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/modeling-submission/modeling-submission.component.spec.ts:468]:
```suggestion
        expect(unreferencedFeedback!).toHaveLength(1);
```

File: src/test/javascript/spec/component/modeling-submission/modeling-submission.component.spec.ts [hunk 4]
```diff
@@ -428,4 +428,47 @@ describe('ModelingSubmission Management Component', () => {
...
+        expect(unreferencedFeedback![0].isSubsequent).toBe(undefined);
+        expect(referencedFeedback).not.toBe(undefined);
+        expect(referencedFeedback!.length).toBe(1);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/modeling-submission/modeling-submission.component.spec.ts:471]:
```suggestion
        expect(referencedFeedback!).toHaveLength(1);
```

File: src/test/javascript/spec/component/file-upload-submission/file-upload-submission.component.spec.ts [hunk 5]
```diff
@@ -287,4 +289,42 @@ describe('FileUploadSubmissionComponent', () => {
...
+            } as Feedback,
+        ];
+
+        comp.result = new Result();
+        comp.result.feedbacks = feedbacks;
+
+        const unreferencedFeedback = comp.unreferencedFeedback;
+
+        expect(unreferencedFeedback).not.toBe(undefined);
+        expect(unreferencedFeedback!.length).toBe(2);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/file-upload-submission/file-upload-submission.component.spec.ts:326]:
```suggestion
        expect(unreferencedFeedback!).toHaveLength(2);
```

File: src/test/javascript/spec/component/text-result/text-result.component.spec.ts [hunk 6]
```diff
@@ -175,4 +175,63 @@ describe('TextResultComponent', () => {
...
+            } as Feedback,
+        ];
+
+        submission.blocks = blocks;
+        result.submission = submission;
+        result.feedbacks = feedback;
+
+        component.result = result;
+
+        expect(component.textResults.length).toBe(2);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/text-result/text-result.component.spec.ts:231]:
```suggestion
        expect(component.textResults).toHaveLength(2);
```

**Review by willergeorg — COMMENTED** [review]
I tested on TS1 the case that the assessment was not shown correctly in the programming exercise editor (subsequent feedback was not colored grey). This worked for me now. Good job!
I also had a look at your code. My comments regarding the translations are only preferences as I found that the enumeration was not clearly separated. But let me know your opinion on this.
Regarding the test code I marked you some small things.

File: src/main/webapp/i18n/en/exercise.json (hunk 2 above)
Comment (imstreetspirit) [reply src/main/webapp/i18n/en/exercise.json:61]: it makes sense 👍

**Review by willergeorg — APPROVED** [review]
Code lgtm now. Functionality worked for me as well.

**Review by mallgaier — APPROVED** [review]
Tested together with @lBruchmann , @Strohgelaender , @VinnavanSembian and @willergeorg .
The functionality was implemented as expected.

**Review by VinnavanSembian — APPROVED** [review]
tested on ts1 with @lBruchmann @Strohgelaender @mallgaier @willergeorg. Everything works as expected now! Thank you!

**Review by lBruchmann — APPROVED** [review]
Tested on ts1. Tested together with @Strohgelaender, @mallgaier, @willergeorg, and @VinnavanSembian. We followed the testing steps. Everything works as expected.

**Review by Strohgelaender — APPROVED** [review]
Tested with @lBruchmann @VinnavanSembian @mallgaier @willergeorg on TS1. Everything worked as expected, thanks for implementing my request.

**Review by MerlinMehmed — APPROVED** [review]
Code changes look good to me!

**Pull request merged**